Allow custom file extensions for directory journal

comkieffer opened the pull request:
This PR builds on https://github.com/jrnl-org/jrnl/pull/1789 (Closes issue #1289). 

In short, it adds a new configuration option `extension` that controls the extension for the files created with a folder journal. 

### Checklist

- [X] I have read the [contributing doc](https://github.com/jrnl-org/jrnl/blob/develop/docs/contributing.md).
- [X] I have included a link to the relevant issue number.
- [X] I have checked to ensure there aren't other open [pull requests](../pulls)
  for the same issue.
- [X] I have written new tests for these changes, as needed.

comkieffer committed:
Allow custom file extensions for directory journal

comkieffer committed:
Test for the creation of all the expected files

The test previously only tested that one of the expected files was
created.

thw26 commented:
Thanks for building on my PR! I have not had the time to return to it since the review.

micahellison commented on tests/bdd/features/file_storage.feature:
```diff
@@ -10,13 +10,21 @@ Feature: Journals iteracting with the file system in a way that users can see
         Then we should get no error
         And the journal directory should contain
             2013/07/23.txt
+            2014/07/23.txt
```
Date doesn't line up with the test case in default US date layout -- either this should be changed to 2014/03/07.txt or the second `we run` line should be changed to: `And we run "jrnl 7/23/2014: Second entry of journal."`

comkieffer replied:
Fixed, I've updated the dates.

micahellison commented on tests/bdd/features/file_storage.feature:
```diff
@@ -10,13 +10,21 @@ Feature: Journals iteracting with the file system in a way that users can see
+    Scenario: Adding entries to a Folder journal with a custom extension should generate date files
+        Given we use the config "empty_folder_with_extension.yaml"
+        When we run "jrnl 23 July 2013: Testing folder journal."
+        Then we should get no error
```
Congratulations, your test uncovered an existing bug! 🎉 I filed it in #1894.

Looks like `empty_folder.yaml` got past this issue by having an "empty" file in it: `tests/data/journals/empty_folder/empty`

Feel free to copy that "empty" file to `tests/data/journals/empty_folder_with_extension/empty` to get this working.

comkieffer replied:
I can't replicate this. I'm not sure what might have caused the issue. 

I re-ran the tests without the `empty` file in `test/data/journals/empty_folder` and the tests still passed.

micahellison replied:
Looks like bug #1894 is Windows-only, and this is combined with a limitation of git: since we can only commit files (not directories), that `empty_folder` doesn't exist on the remote repository until we push a file in it.

micahellison reviewed (changes requested):
Sorry for taking so long to get to this PR, but also, thank you for the excellent PR. I really appreciate that you're modifying documentation, adding tests, and adding thoughtful changes within the norms of the existing code.

The GitHub Actions tests were broken when your PR was first submitted, but they should be workable now if you merge in develop. Otherwise, I added notes about a couple small tasks that should get these new tests passing.

micahellison committed:
Skip Windows due to mysterious os.makedirs failure

wren committed:
Fix handling for complex extensions

This also handles additional input for extensions in case the user
isn't fully aware of how extensions work

micahellison commented on jrnl/journals/FolderJournal.py:
```diff
@@ -143,15 +148,16 @@ def _get_month_folders(path: pathlib.Path) -> list[pathlib.Path]:
-    def _get_day_files(path: pathlib.Path) -> list[str]:
-        for child in path.glob(DAY_PATTERN):
+    def _get_day_files(path: pathlib.Path, extension: str) -> list[str]:
+        for child in path.glob(DAY_PATTERN + extension):
```
I think the only thing left on this PR is supporting a lack of dots in the extension. For instance, I'd expect this to work the same whether my config value is `extension: md` or `extension: .md`
